### PR TITLE
[InstanceChoice] Add the instance choice specialization pass to firtool

### PR DESCRIPTION
**Pull request by prithayan**
This PR adds the instance choice specialization pass to the `firtool` pipeline.
Add optional arguments to `firtool` to select the specialization option for the instance choice.
Then update the pass to use the default module, when no options are specified.
Update lit tests to check for valid instance choice lowering.

### CODE REVIEW

File: lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp [hunk 1]
```diff
@@ -68,33 +64,46 @@ struct SpecializeOptionPass
...
 
+    bool analysisPreserved = numInstances == 0;
+    circuit->walk([&](OptionOp optionOp) {
+      optionOp->erase();
+      analysisPreserved = false;
+    });
+    if (analysisPreserved)
+      markAllAnalysesPreserved();
+
     if (failed)
```
Comment (uenoku) [inline lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp:97]: failed is dead now?

File: include/circt/Firtool/Firtool.h
Comment (seldridge) [inline include/circt/Firtool/Firtool.h:1]: Please have the PR reflect any changes to this file in the corresponding area of the C-API.

File: include/circt/Dialect/FIRRTL/Passes.h [hunk 2]
```diff
@@ -202,7 +202,8 @@ std::unique_ptr<mlir::Pass> createProbesToSignalsPass();
 
 std::unique_ptr<mlir::Pass> createSpecializeLayersPass();
 
-std::unique_ptr<mlir::Pass> createSpecializeOptionPass();
+std::unique_ptr<mlir::Pass> createSpecializeOptionPass(
```
Comment (uenoku) [inline include/circt/Dialect/FIRRTL/Passes.h:205]: I think we can remove `std::optional`. You can use empty array as `nullopt`.

File: lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp [hunk 3]
```diff
@@ -68,33 +64,46 @@ struct SpecializeOptionPass
         &getContext(), circuit.getOps<FModuleOp>(), [&](auto module) {
           module.walk([&](firrtl::InstanceChoiceOp inst) {
             auto it = selected.find(inst.getOptionNameAttr());
+            FlatSymbolRefAttr target;
             if (it == selected.end()) {
-              inst.emitError("missing specialization for option ")
+              target = inst.getDefaultTargetAttr();
+              inst->emitWarning("missing specialization for option ")
                   << inst.getOptionNameAttr();
```
Comment (uenoku) [inline lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp:71]: Do we want to keep this warning if we decided to use default lowering?

File: lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp [hunk 4]
```diff
@@ -68,33 +64,46 @@ struct SpecializeOptionPass
...
-              failed = true;
-              return;
-            }
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp:76]: I realize that this all-or-nothing behavior is originally how this worked.

However, it seems odd to force a user to specialize everything or nothing as opposed to just specializing what the user requested.

**Review by uenoku — APPROVED** [review]
LGTM, please update capi as well.

File: test/firtool/instchoice.fir [hunk 5]
```diff
@@ -26,6 +28,8 @@ circuit Foo:
     ; CHECK-SAME:   @FPGA -> @FPGATarget,
     ; CHECK-SAME:   @ASIC -> @ASICTarget
     ; CHECK-SAME: } (in clock: !firrtl.clock)
+    ; ASIC: hw.instance "inst" @ASICTarget(clock: %clock: !seq.clock) -> ()
+    ; DEFAULT: hw.instance "inst" @DefaultTarget(clock: %clock: !seq.clock) -> ()
```
Comment (seldridge) [inline test/firtool/instchoice.fir:32]: 🤔 Making the default behavior be "replace instance choices with the default" also seems slightly odd. It seems like the default should be "do nothing". I realize that instance choice is unusable right now unless specialization of every option is run, though.

**Review by seldridge — COMMENTED** [review]
The roughly sister pass of this is `SpecializeLayers` and that went a slightly different direction than adding an option. Instead that puts the information directly onto the `CircuitOp` via a mandatory attribute.

While I can't recall the exact reasons for this design point, consistency between the two approaches may be good.

@youngar or @rwy7: Do you have any thoughts about the above?

File: test/firtool/instchoice.fir (hunk 5 above)
Comment (prithayan) [reply test/firtool/instchoice.fir:32]: Maybe it depends on how we define it, the default target, for an instance choice can be defined as the target when no option is specified. Also, if nothing is done here, when should the instance choice be handled in the pipeline ?

File: lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp (hunk 4 above)
Comment (prithayan) [reply lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp:76]: In the current implementation, the user need not specialize everything, it would be lowered to the default target, if nothing is specified for an instance choice. The user can specify only the choices that are relevant, and the rest can be lowered to the default target.

File: test/firtool/instchoice.fir (hunk 5 above)
Comment (seldridge) [reply test/firtool/instchoice.fir:32]: If no default is specified, then the instance choice gets lowered to how the FIRRTL ABI defines this lowering. (This is not specified right now. However, I have one such ABI PR'd here: https://github.com/chipsalliance/firrtl-spec/pull/155.)

The underlying problem is that instance choice was only handled for this narrow, internal use case of CIRCT choosing a value for all instance choices as opposed to the more general user request for Verilog elaboration-time selection. 🥲 

The motivation here is more about aligning this with how layers are enabled and disabled, e.g., `--enable-layer=Verification` specializes away the `Verification` layer, but it leaves any other layers in place. It would seem better to have instance choice specialization work the same way, e.g., `--specialize-instance-choice=Performance=ASIC` would not affect any other instance choice in the circuit.

We do want to move in this direction. However, we don't have to move there right now with this PR.

WDYT?

Comment (davidbiancolin) [reply test/firtool/instchoice.fir:32]: I think this makes more sense, but i'm assuming you meant to write: 

`--specialize-instance-choice=Platform=ASIC`

To specify which target is being specialized? 

Also, not to derail this (and i see there was some discussion of this in the spec PR) but it is useful to be able to specialize different instances of the same `instance_choice` to write miter-style circuits for testing the choices themselves (esp. when they are used in part of a larger circuit under test). I guess here it might just be fine to do the specialization at VCS-compile time though. For this reason i also like that default be explicitly named (as also discussed in the spec PR).

Comment (davidbiancolin) [reply test/firtool/instchoice.fir:32]: So in the short run, we'd need a flag to achieve the effect of specializing all modules to their default and just provide that when we're passing in circuits that contain instance choices for now?

Comment (seldridge) [reply test/firtool/instchoice.fir:32]: > I think this makes more sense, but i'm assuming you meant to write: 
> 
> `--specialize-instance-choice=Platform=ASIC`
> 
> To specify which target is being specialized? 

Yes!

> it is useful to be able to specialize different instances of the same instance_choice to write miter-style circuits for testing the choices themselves (esp. when they are used in part of a larger circuit under test)

Per-instance specialization currently works. It does create a blow-up in the number of options. This is an intentional choice that complicates the design. E.g., consider instance choice used in a multicore design to replace a core with a model. The designer needs to decide what swapping flexibility they want to have. If they want to have complete flexibility to swap every core arbitrarily in an n-core design, that would have to be n-different options. They likely don't want that and will choose to have one instance choice which swaps all of the cores except one.

I'd prefer to not support arbitrary per-instance after-the-fact specialization as that amounts to mandated module duplication which then has interactions with must-dedup and designer expectations around deduplication. I'm more comfortable with this if CIRCT-mandated deduplication (must-dedup) is removed.

> So in the short run, we'd need a flag to achieve the effect of specializing all modules to their default and just provide that when we're passing in circuits that contain instance choices for now?

I'm good with either:

1. All instance choice specializations are specified up front. 
2. There is a catch-all `-specialize-to-default-if-no-override` which will, after applying selected choices, will then convert everything to the default choice.

Comment (davidbiancolin) [reply test/firtool/instchoice.fir:32]: After more thought, i think strictness of (1) is probably desirable to avoid accidental footguns? We can always add the flag later. I'll defer to you and @prithayan on this one.

Comment (prithayan) [reply test/firtool/instchoice.fir]: I've decided to go with the option 1, unless all the instance choice specializations are specified, the pass will error out. 
Once the lowering of instance choice through the pipeline to verilog is supported, we can get rid of the error.

File: include/circt/Firtool/Firtool.h
Comment (prithayan) [reply include/circt/Firtool/Firtool.h:1]: I'm trying to figure out the corresponding type for `ArrayRef` in C-API, any idea ?

Comment (seldridge) [reply include/circt/Firtool/Firtool.h:1]: I don't see any uses of `ArrayRef` for the C-API. This may need to get funneled through an `ArrayAttr`.

File: lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp [hunk 6]
```diff
@@ -89,12 +85,24 @@ struct SpecializeOptionPass
           });
         });
 
+    bool analysisPreserved = numInstances == 0;
+    circuit->walk([&](OptionOp optionOp) {
+      optionOp->erase();
```
Comment (prithayan) [inline lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp:90]: @seldridge , @uenoku , just wanted to make sure, this makes sense right ?
 The `OptionOp`  wouldn't be useful for any other passes downstream, even if we ignore an `InstanceChoiceOp`, when it is not specialized, and left in IR after this pass.

File: include/circt/Dialect/FIRRTL/Passes.h (hunk 2 above)
Comment (prithayan) [reply include/circt/Dialect/FIRRTL/Passes.h]: Fixed.

File: lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp (hunk 6 above)
Comment (seldridge) [reply lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp:90]: This would only be safe for options that were specialized. If any `InstanceChoiceOp` is left in, e.g., if the circuit is partially instance-choice specialized, then this would create an invalid symbol reference.

Comment (prithayan) [reply lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp:91]: Yes, that makes sense. But since, there are no other passes that can handle the `OptionOp` maybe its okay to delete them. And this pass also errors out on partially specified instance choice. WDYT ?

**Comment by prithayan** [comment]
PR is ready for another review, @uenoku , @seldridge please take a look if you get a chance.

- I updated the select option to follow the specialize layers style.
- Removed the lowering to default and instead error out on partially specified choice.

File: lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp (hunk 6 above)
Comment (seldridge) [reply lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp:92]: All the other passes are _supposed_ to handle these things, just it's not implemented, yet. 🫠 The concern is more about if this pass is guaranteed to specialize away everything (which we eventually don't want), then it's fine. However, if it is going to leave anything in, then it shouldn't delete everything.

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 7]
```diff
@@ -5825,10 +5825,24 @@ ParseResult FIRCircuitParser::parseCircuit(
...
 
+  auto arrayToAttr = [&](const ArrayRef<std::string> strArray) {
+    SmallVector<StringRef> attr;
+    for (const auto &str : strArray)
+      attr.push_back(str);
+    OpBuilder builder(getContext());
+    if (attr.empty())
+      return ArrayAttr();
+    return builder.getStrArrayAttr(attr);
+  };
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:5836]: This is super close to `parseLayers`. Can these be combined?

File: lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp [hunk 8]
```diff
@@ -27,40 +27,38 @@ namespace {
...
-    for (const auto &optionAndCase : select) {
-      size_t eq = optionAndCase.find("=");
-      if (eq == std::string::npos) {
-        mlir::emitError(circuit.getLoc(),
-                        "invalid option format: \"" + optionAndCase + '"');
-        return signalPassFailure();
-      }
+    if (auto choiceAttr = circuit.getSelectInstChoiceAttr()) {
+      for (auto attr : choiceAttr.getAsRange<StringAttr>()) {
+        const auto &optionAndCase = attr.getValue().str();
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp:37]: Is a `const std::string &` safe to take for a `.str()` that isn't given a temporary?

File: lib/Firtool/Firtool.cpp [hunk 9]
```diff
@@ -55,6 +55,7 @@ LogicalResult firtool::populatePreprocessTransforms(mlir::PassManager &pm,
 LogicalResult firtool::populateCHIRRTLToLowFIRRTL(mlir::PassManager &pm,
                                                   const FirtoolOptions &opt,
                                                   StringRef inputFilename) {
+
```
Comment (seldridge) [inline lib/Firtool/Firtool.cpp:58]: Mega-nit: unrelated whitespace change.

File: lib/Firtool/Firtool.cpp [hunk 10]
```diff
@@ -133,6 +134,7 @@ LogicalResult firtool::populateCHIRRTLToLowFIRRTL(mlir::PassManager &pm,
 
   // Must run this pass after all diagnostic passes have run, otherwise it can
   // hide errors.
```
Comment (seldridge) [inline lib/Firtool/Firtool.cpp:136]: Mega-nit: can you update the comment to indicate that this is plural?

File: test/Dialect/FIRRTL/specialize-option-errors.mlir
Comment (seldridge) [inline test/Dialect/FIRRTL/specialize-option-errors.mlir:1]: Can this PR include a test of the error if the specialization is not fully specified?

**Review by seldridge — APPROVED** [review]
Nice. Some comments/nits throughout. 

Can the PR add a test of the error behavior if the specialization is not fully specified?

File: lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp (hunk 6 above)
Comment (prithayan) [reply lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp:92]: Right, for now, this pass currently doesn't leave any `InstanceChoiceOp`, it should error out if a choice is not selected. So, it should be fine to delete it.
But the first pass that needs to handle `OptionOp` is `LowerToHW`, there is no `HW` counterpart for it.

File: test/firtool/instchoice.fir (hunk 5 above)
Comment (davidbiancolin) [reply test/firtool/instchoice.fir]: So if you went with (1) how do you specialize to `Default`?

File: lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp (hunk 6 above)
Comment (seldridge) [reply lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp:92]: Yep... just showing something end-to-end would be cool. FWIW, I am totally fine with `LowerToHW` erroring with "unsupported operation". That's all part of getting a feature turned on incrementally. 😉

File: test/Dialect/FIRRTL/specialize-option-errors.mlir
Comment (prithayan) [reply test/Dialect/FIRRTL/specialize-option-errors.mlir:1]: Added the test.

File: lib/Firtool/Firtool.cpp (hunk 10 above)
Comment (prithayan) [reply lib/Firtool/Firtool.cpp]: Updated the comment.

File: lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp (hunk 8 above)
Comment (prithayan) [reply lib/Dialect/FIRRTL/Transforms/SpecializeOption.cpp]: Thanks for spotting this. Fixed.

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp (hunk 7 above)
Comment (prithayan) [reply lib/Dialect/FIRRTL/Import/FIRParser.cpp]: Merged them.

**Comment by prithayan** [comment]
Thanks for the review and discussion everyone. I'm planning to split this into a separate PR for the parser changes and will merge them as two commits.

**Comment by seldridge** [comment]
> Thanks for the review and discussion everyone. I'm planning to split this into a separate PR for the parser changes and will merge them as two commits.

👍 Always feel empowered to slice things up into smaller commit chunks. 💯
